rllib/env/multi_agent_env_runner.py

```diff
@@ -0,0 +1,887 @@
+            # if automated).
+            self._episode.add_env_reset(observations=obs, infos=infos)
+
+            # Get the states for all agents.
+            states = initial_states
+        # Do not reset environments, but instead continue in already started episodes.
+        else:
+            # Pick up stored observations from previous timesteps.
+            obs = self._episode.get_observations(as_list=True)[0]
+            # Get the states from the buffer or get the initial states.
+            # TODO (simon): Do we need to iterate here over agents? Or can
+            # one agent have no buffered state while another one has?
+            states = {
+                agent_id: initial_states[agent_id]
+                if agent_state is None
+                else agent_state
+                for agent_id, agent_state in self._states.items()
+            }
+
+        # Loop through timesteps.
+        env_steps = 0
+        agent_steps = 0
+        ts = 0
+
+        while ts < num_timesteps:
+            # Act randomly.
+            if random_actions:
+                # Note, to get sampled actions from all agents' action
+                # spaces we need to call `MultiAgentEnv.action_space_sample()`.
+                if self.env.unwrapped._action_space_iin_preferred_format:
+                    actions = self.env.action_space.sample()
+                # Otherwise, `action_space_sample()` needs to be implemented.
+                else:
+                    actions = self.env.action_space_sample()
+                # Remove all actions for agents that had no observation.
+                actions = {
+                    agent_id: agent_action
+                    for agent_id, agent_action in actions.items()
+                    if agent_id in obs
+                }
+
+            else:
+                # TODO (simon): This is not correct `forward()` expects
+                # `SampleBatchType`.
+                # Note, `RLModule`'s `forward()` methods expect `NestedDict`s.
+                # Note, we only consider for states and obs agents that step.
+                batch: MultiAgentDict = {
+                    self._agent_id_to_module_id[agent_id]: {
+                        STATE_IN: tree.map_structure(
+                            lambda s: self._convert_from_numpy(s),
+                            states[self._agent_id_to_module_id[agent_id]],
+                        ),
+                        SampleBatch.OBS: self._convert_from_numpy(
+                            np.expand_dims(agent_obs, axis=0)
+                        ),
+                    }
+                    for agent_id, agent_obs in obs.items()
+                }
+                # TODO (Sven, Simon): The `RLModule` has `SampleBatchType` as input
+                # type. Only the _forward_x()` methods have a `NestedDict`. Shall we
+                # compile to `SampleBatchType` here and in `SingleAgentEnvRunner`?
```

Comment:
I don't think, we need to compile a SampleBatch.
I'm not even a fan of NestedDict anymore. Let's try to keep it as simple dicts instead as long as possible.